scripts/monitoring/monitoring.py

```diff
@@ -0,0 +1,536 @@
+    return global_config
+
+
+@click.command
+@click.option(
+    "--config",
+    "-c",
+    "config_path",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, path_type=Path),
+    metavar="PATH",
+    help="Config file with chains configuration",
+)
+def main(
+    config_path: Path,
+) -> None:
+    setup_logging(log_level="DEBUG", log_json=False)
+
+    set_config(config_path)
+
+    transfers = create_transfers_object(fetch_events())
+
+    cleanup_transfers(transfers)
+
+    ctx = Context()
+    ctx.agent_address = get_config()["agent_address"]
+
+    analyze_transfers(transfers, ctx)
+
+
+def cleanup_transfers(transfers: TransfersMap) -> None:
+    for requestId, values in transfers.copy().items():
+        if values.get("created") is None:
+            del transfers[requestId]
+
+        claims = values.get("claimed")
+        if claims is not None:
+            for events in claims.values():
+                events.sort(key=lambda a: -1 if a.last_challenger == ADDRESS_ZERO else 1)
+
+
+def fetch_events() -> ChainToEvent:
+    deployment_info = beamer.contracts.load_deployment_info(get_config()["deployment_dir"])
+    events = {}
+
+    for chain_id, (rpc, start_block) in get_config()["rpcs"].items():
+        web3 = make_web3(rpc)
+        assert chain_id == ChainId(web3.eth.chain_id)
+
+        info = deployment_info[ChainId(chain_id)]
+        contracts = beamer.contracts.make_contracts(web3, info)
+
+        ef = beamer.events.EventFetcher(
+            web3, (contracts["RequestManager"], contracts["FillManager"]), start_block
+        )
+
+        events[chain_id] = ef.fetch()
+
+    return cast(ChainToEvent, events)
+
+
+def make_web3(rpc: str) -> Web3:
+    web3 = Web3(Web3.HTTPProvider(rpc, request_kwargs=dict(timeout=5)))
+    web3.middleware_onion.inject(geth_poa_middleware, layer=0)
+    return web3
+
+
+def get_agent_liquidity(
+    agent_address: str, rpcs: dict, tokens: TokensConfig
+) -> dict[str, dict[ChainId, int]]:
+    contract_abi = load_ERC20_abi()
+    liquidity: dict = defaultdict(dict)
+    agent_address = Web3.toChecksumAddress(agent_address)
+
+    for name, chain_to_token_mapping in tokens.items():
+        for chain_id, token_address in chain_to_token_mapping:
+            contract_address = Web3.toChecksumAddress(token_address)
+            web3 = make_web3(rpcs[int(chain_id)][0])
+            contract = web3.eth.contract(address=contract_address, abi=contract_abi)
+            token_decimals = contract.functions.decimals().call()
+            liquidity[name][chain_id] = (
+                contract.functions.balanceOf(agent_address).call() * 1 * 10**-token_decimals
+            )
+
+    for chain_id, (rpc, _start_block) in rpcs.items():
+        web3 = make_web3(rpc)
+        agent_address = Web3.toChecksumAddress(agent_address)
+        assert ChainId(int(chain_id)) == ChainId(web3.eth.chain_id)
+        liquidity["eth"][chain_id] = web3.eth.get_balance(agent_address) * 1 * 10**-18
```

Comment:
are we repeating ourselves here ? the previous loops is kind of setting funds imo